Introduce login page as homepage

aglinxinyuan opened the pull request:
**Motivation:**
Currently, the user stays on the same page after they log out. Each page needs to track user change events, which is inconvenient and insecure.

**Solution:**
In this PR, I introduce a new login page, and users have to log in to view any system content and will be redirected to the login page after logout.
I also introduced an auth-guard login filter, which can be reused for the permission system.
The login page is responsive and will work with different window sizes.

I also removed the original login modal.

![image](https://user-images.githubusercontent.com/11544314/206892455-b0ee048e-8c4a-4941-8e70-137c96568882.png)

Xiao-zhen-Liu commented:
Discussion 2022.12.07: Work in progress.

Yicong-Huang commented on core/new-gui/src/app/dashboard/component/login/login.component.ts:
```diff
@@ -0,0 +1,133 @@
+  templateUrl: "./login.component.html",
+  styleUrls: ["./login.component.scss"],
+})
+export class LoginComponent implements OnInit {
```
it seems we are doing many things other than login on this component, for example we are showing some information for texera (blog, wiki), showing option for people to sign up, etc.

I think `LoginComponent` does not describe this page accurately. This component is in parallel with `WorkspaceComponent` and `DashboardComponent`, shall we name it as `LandingComponent`, where it has child components to do login, registration, information display, etc?

Yicong-Huang replied:
Or `HomeComponent`?

aglinxinyuan replied:
HomeComponent is good.

Yicong-Huang commented on core/new-gui/src/environments/environment.default.ts:
```diff
@@ -36,12 +36,12 @@ export const defaultEnvironment = {
    */
-  userSystemEnabled: false,
+  userSystemEnabled: true,
 
   /**
    * whether local login is enabled
    */
-  localLogin: true,
+  localLogin: false,
```
shall we revert the default value changes?

aglinxinyuan replied:
I pushed this by mistake.

Yicong-Huang commented on core/new-gui/src/app/dashboard/service/auth-guard/auth-guard.service.ts:
```diff
@@ -0,0 +1,14 @@
+import { Router, CanActivate } from "@angular/router";
+import { UserService } from "../../../common/service/user/user.service";
+@Injectable()
+export class AuthGuardService implements CanActivate {
```
it seems that this service is not only used by dashboard components but also workspace components (i.e., if logout from anywhere, it should be redirected back to login). 

Shall we put this service under `src/app/common/service/` path?

aglinxinyuan replied:
Yes.

Yicong-Huang commented on core/new-gui/src/app/dashboard/service/auth-guard/auth-guard.service.ts:
```diff
@@ -0,0 +1,14 @@
+import { Router, CanActivate } from "@angular/router";
+import { UserService } from "../../../common/service/user/user.service";
+@Injectable()
+export class AuthGuardService implements CanActivate {
```
Please add a class block comment to describe this service.  You can refer to other services' documentation.

aglinxinyuan replied:
commented

Yicong-Huang commented on core/new-gui/src/app/dashboard/component/login/login.component.ts:
```diff
@@ -0,0 +1,133 @@
+  };
+
+  /**
+   * This method is respond for the sign in button in the pop up
+   * It will send data inside the text entry to the user service to login
```
please fix the grammar here.  also, are we still using pop up? if not please update the comments.

aglinxinyuan replied:
fixed.

Yicong-Huang commented on core/new-gui/src/app/dashboard/component/login/login.component.ts:
```diff
@@ -0,0 +1,133 @@
+      .pipe(filter(isDefined))
+      .pipe(untilDestroyed(this))
+      .subscribe(
+        Zone.current.wrap(() => {
+          this.router.navigate(["dashboard/workflow"]);
+        }, "")
```
do we still need the Zone wrapper?

Yicong-Huang commented on core/new-gui/src/app/common/service/user/user.service.ts:
```diff
@@ -46,7 +47,10 @@ export class UserService {
 
   public logout(): void {
-    this.authService.logout().subscribe(_ => this.changeUser(undefined));
+    this.authService.logout().subscribe(_ => {
+      this.changeUser(undefined);
+      this.router.navigate(["login"]);
```
we should not do re-routing inside `user-serivce`. The scope of user service is clear: it manages the current user information of the system, and navigation is NOT its task.

There should be another service (or component, if applicable) to listen to and handle `userChanged` event, so that it can navigate accordingly.

Yicong-Huang reviewed (changes requested):
I've played with the new page, it looks really good! thanks for the effort!

Some general suggestions:
1. Name the components/variables/services precisely. They should do what their names described.
2. Make sure you update the comments and docstrings accordingly when you move things around. a wrong comment is worse than no comment.
3. Avoid unnecessary changes. 
4. Add a test case for the new component(s), just to testify if it(they) can be initialized is enough for now.

Please refer to the code to see detailed comments.

aglinxinyuan commented on core/amber/src/main/resources/application.conf:
```diff
@@ -56,7 +56,7 @@ cache {
 }
 
 user-sys {
-    enabled = false
```
This one is mistake. Going to fix.

aglinxinyuan commented on core/new-gui/src/app/dashboard/component/login/login.component.scss:
```diff
@@ -0,0 +1,51 @@
+  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans",
+    "Droid Sans", "Helvetica Neue", sans-serif;
+
+  background-image: url(data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMTgiIGhlaWdodD0iMTgiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyI+PGcgZmlsbD0ibm9uZSIgZmlsbC1ydWxlPSJldmVub2RkIj48cGF0aCBkPSJNMTcuNiA5LjJsLS4xLTEuOEg5djMuNGg0LjhDMTMuNiAxMiAxMyAxMyAxMiAxMy42djIuMmgzYTguOCA4LjggMCAwIDAgMi42LTYuNnoiIGZpbGw9IiM0Mjg1RjQiIGZpbGwtcnVsZT0ibm9uemVybyIvPjxwYXRoIGQ9Ik05IDE4YzIuNCAwIDQuNS0uOCA2LTIuMmwtMy0yL ...
```
This is google logo.

Yicong-Huang replied:
then nvm. shallow keep a comment so people know this is the google logo?

aglinxinyuan replied:
It's inside google button css, but I can comment it anyway.

aglinxinyuan commented:
> I've played with the new page, it looks really good! thanks for the effort!
> 
> Some general suggestions:
> 
> 1. Name the components/variables/services precisely. They should do what their names described.
> 2. Make sure you update the comments and docstrings accordingly when you move things around. a wrong comment is worse than no comment.
> 3. Avoid unnecessary changes.
> 4. Add a test case for the new component(s), just to testify if it(they) can be initialized is enough for now.
> 
> Please refer to the code to see detailed comments.

Thanks for the advice.
Changes were made accrodingly.
1. Components renamed.
2. Comments and docstrings added and updated.
3. Removed unnecessary changes.
4. Test cases added.

Yicong-Huang commented on core/new-gui/src/app/home/component/home.component.html:
```diff
@@ -0,0 +1,54 @@
+        <br />
+      </p>
+
+      <h4>
+        <a href="https://youtu.be/2gfPUZNsoBs">Demo Video</a>
```
I wonder if we are also interested in including the GitHub repo link here. Ideally to show it as a badge. 

Something like [this](https://ghbtns.com):
<img width="694" alt="Screenshot 2022-12-10 at 21 12 52" src="https://user-images.githubusercontent.com/17627829/206887654-888c470b-941e-4881-a105-729f00920953.png">

Yicong-Huang reviewed (approved):
LGTM. Left minor comments.
